components/debounce/browser/debounce_service_impl.cc

```diff
@@ -0,0 +1,49 @@
+/* Copyright (c) 2021 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/debounce/browser/debounce_service_impl.h"
+
+#include <memory>
+#include <vector>
+
+#include "base/logging.h"
+#include "brave/components/debounce/browser/debounce_download_service.h"
+
+namespace debounce {
+
+DebounceServiceImpl::DebounceServiceImpl(
+    DebounceDownloadService* download_service)
+    : download_service_(download_service), weak_factory_(this) {}
+
+DebounceServiceImpl::~DebounceServiceImpl() {}
+
+bool DebounceServiceImpl::Debounce(
+    const GURL& original_url,
+    const net::SiteForCookies& original_site_for_cookies,
+    GURL* final_url) {
+  bool changed = false;
+  GURL current_url = original_url;
+  net::SiteForCookies current_site_for_cookies = original_site_for_cookies;
+  std::vector<std::unique_ptr<DebounceRule>>* rules =
+      download_service_->rules();
+
+  // Debounce rules are applied in order. All rules are checked on every URL. If
```

Comment:
how many rules are there? I think we need a more efficient lookup mechanism if there is anything more than a trivial number of them

Reply:
i'd say we can start with adding an `SCOPED_UMA_HISTOGRAM_TIMER` to measure with real-world rules.

Reply:
Added a host_cache in https://github.com/brave/brave-core/commit/4aa1a0bf4d75ecc201119e062fdd1e55ec6e6ec4 as discussed in Slack